### PR TITLE
Added qt_sink_color class

### PR DESCRIPTION
**Pull request by mguludag**
Hello, I added qt_sink_color_mt/st that uses bundled fmtlib's color and text style syntax, also implemented set_color function for we can able to change color or style later.
set_color function example:
`set_color(spdlog::level::err, fmt::fg(fmt::color::white) | fmt::bg(fmt::color::red) | fmt::emphasis::bold | fmt::emphasis::italic);`

### CODE REVIEW

**Comment by gabime** [comment]
Thanks. Since spdlog now can be compiled without fmt ( `#define SPDLOG_USE_STD_FORMAT` ) this is a little problematic to rely on fmt color. Also, is some installations an external fmt is used and the fmt/bundled/color.h doesnt exist at all.

**Comment by mguludag** [comment]
> Thanks. Since spdlog now can be compiled without fmt ( `#define SPDLOG_USE_STD_FORMAT` ) this is a little problematic to rely on fmt color. Also, is some installations an external fmt is used and the fmt/bundled/color.h doesnt exist at all.

I forgot this scenarios, I added guards for this

File: include/spdlog/sinks/qt_sinks.h [hunk 1]
```diff
@@ -4,46 +4,55 @@
...
 // qt_sink class
 //
 namespace spdlog {
 namespace sinks {
-template<typename Mutex>
-class qt_sink : public base_sink<Mutex>
-{
+#if !defined(SPDLOG_USE_STD_FORMAT)
+template <typename Mutex>
+class qt_sink_color;
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:34]: why the forward declaration is needed?

File: include/spdlog/sinks/qt_sinks.h [hunk 2]
```diff
@@ -53,50 +62,184 @@ class qt_sink : public base_sink<Mutex>
     std::string meta_method_;
 };
 
+#if !defined(SPDLOG_USE_STD_FORMAT)
+template <typename Mutex>
+class qt_sink_color : public qt_sink<Mutex> {
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:67]: I don't see much benefit in inherit from qt_sink here.

File: include/spdlog/sinks/qt_sinks.h [hunk 3]
```diff
@@ -53,50 +62,184 @@ class qt_sink : public base_sink<Mutex>
...
+        auto bg = fmt::detail::make_background_color<fmt::detail::char8_type>(color.get_background());
+        auto fg = fmt::detail::make_foreground_color<fmt::detail::char8_type>(color.get_foreground());
+        auto em = fmt::detail::make_emphasis<fmt::detail::char8_type>(color.get_emphasis());
+
+        str.append(bg.begin(), bg.end());
+        str.append(fg.begin(), fg.end());
+        str.append(em.begin(), em.end());
+
+        colors_[color_level] = str;
+    }
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:99]: str can be moved here to avoid the copy

File: include/spdlog/sinks/qt_sinks.h [hunk 4]
```diff
@@ -4,46 +4,55 @@
...
-class qt_sink : public base_sink<Mutex>
-{
+#if !defined(SPDLOG_USE_STD_FORMAT)
+template <typename Mutex>
+class qt_sink_color;
+#endif
+template <typename Mutex>
+class qt_sink : public base_sink<Mutex> {
+#if !defined(SPDLOG_USE_STD_FORMAT)
+    friend class qt_sink_color<Mutex>;
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:39]: why friend not needed?

File: include/spdlog/sinks/qt_sinks.h [hunk 5]
```diff
@@ -53,50 +62,184 @@ class qt_sink : public base_sink<Mutex>
...
 
+#if !defined(SPDLOG_USE_STD_FORMAT)
+template <typename Mutex>
+class qt_sink_color : public qt_sink<Mutex> {
+public:
+    qt_sink_color(QObject *qt_object = nullptr, bool color_ = true, const std::string &meta_method = "")
+    {
+        qt_sink<Mutex>::qt_object_ = qt_object;
+        qt_sink<Mutex>::meta_method_ = meta_method;
+        should_do_colors_ = color_;
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:73]: Whats the point of shoild_do_colors? If colors not needed this sinks should not be used anyway

File: include/spdlog/sinks/qt_sinks.h [hunk 6]
```diff
@@ -53,50 +62,184 @@ class qt_sink : public base_sink<Mutex>
...
+    {
+        msg.color_range_start = 0;
+        msg.color_range_end = 0;
+        memory_buf_t formatted;
+        base_sink<Mutex>::formatter_->format(msg, formatted);
+
+        if (should_do_colors_ && msg.color_range_end > msg.color_range_start)
+        {
+            // before color range
+            std::string str;
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:123]: str could be made a member and reused to avoid allocations

File: include/spdlog/sinks/qt_sinks.h [hunk 7]
```diff
@@ -53,50 +62,184 @@ class qt_sink : public base_sink<Mutex>
...
+            auto ccode = to_ccode_(colors_[msg.level]);
+            auto reset = to_ccode_(reset_);
+            str.reserve(formatted.size() + ccode.size());
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:126]: reserve should take into account the reset size as well I think.

File: include/spdlog/sinks/qt_sinks.h [hunk 8]
```diff
@@ -53,50 +62,184 @@ class qt_sink : public base_sink<Mutex>
...
-{
+template <typename Factory = spdlog::synchronous_factory>
+inline std::shared_ptr<logger>
+qt_logger_st(const std::string &logger_name, QObject* qt_object, const std::string &meta_method) {
     return Factory::template create<sinks::qt_sink_st>(logger_name, qt_object, meta_method);
 }
+
+//
+// Factory functions (color)
+//
```
Comment (gabime) [inline include/spdlog/sinks/qt_sinks.h:207]: I dont like all those factory functions. too much code. They exist only in the most popular sinks

**Review by gabime — CHANGES_REQUESTED** [review]

File: include/spdlog/sinks/qt_sinks.h (hunk 2 above)
Comment (mguludag) [reply include/spdlog/sinks/qt_sinks.h]: yes, not too much now I changed to base base_sink

File: include/spdlog/sinks/qt_sinks.h (hunk 3 above)
Comment (mguludag) [reply include/spdlog/sinks/qt_sinks.h:78]: Ok, added move semantics

File: include/spdlog/sinks/qt_sinks.h (hunk 4 above)
Comment (mguludag) [reply include/spdlog/sinks/qt_sinks.h]: I want to access qt_sink's priv members from qt_sink_color class

File: include/spdlog/sinks/qt_sinks.h (hunk 5 above)
Comment (mguludag) [reply include/spdlog/sinks/qt_sinks.h]: Ok, I removed this

**Comment by gabime** [comment]
@mguludag How do the colors work with qt? Do the qt widgets understand ansi colors codes?

**Comment by gabime** [comment]
Not sure I understand. In this pr you are using fmt::color. How does it work with qt?

**Comment by mguludag** [comment]
> Not sure I understand. In this pr you are using fmt::color. How does it work with qt?

fmt::color turns into ansi color codes in QString then it goes to slot that includes ansicolor parser, it parse ansicolor codes and convert into QColor then push into FormattedText object with corresponding text part. setting TextCursor's format coming from this FormattedText object and inserting the text from FormattedText with this format into the widgets.

```C++
void AnsiEscapeCodeHandler::appendAnsiText(const QString &text)
{
    auto const formattedText = parseText(FormattedText(text, m_textCursor.charFormat())); // returns QList<FormattedText>

    for (auto const& text_ : qAsConst(formattedText)) {
        m_textCursor.movePosition(QTextCursor::End);
        m_textCursor.setCharFormat(text_.format);
        m_textCursor.insertText(text_.text);
        m_textCursor.movePosition(QTextCursor::End);
    }

}
```

image

**Comment by gabime** [comment]
So it only usable with your FormattedText code? I suggest adding this to your repo instead. There is no point adding to this repo.

**Comment by mguludag** [comment]
> So it only usable with your FormattedText code? I suggest adding this to your repo instead. There is no point adding to this repo.

Unfortunately Qt doesn't show colored text from ansicolors out of the box. Users have to write parser their own or use prewritten codes like qt-creator's source code.

**Pull request closed without merging**